marconi-sidechain/doc/adr/01-sdd-computation.adoc

```diff
@@ -0,0 +1,203 @@
+
+* `marconi-sidechain` will be run by a Stake Pool Operator (SPO)
+* `marconi-sidechain` will communicate with a relay node (not the block producing node)
+
+The main architectural goal we are trying to achieve is to reduce as much as possible the hardware resource usage of a SPO.
+Therefore, indexing the SDD per epoch should be as resource efficient as possible.
+
+== Decision(s)
+
+TBD
+
+== Possible solutions
+
+We will describe all the potential and discussed solutions to index the SDD of each epoch in the following sections.
+
+As a summary, we provide the following caracteristics table and indicate how each solution responds to them.
+This assessment is measured in regards to two components: Cardano Node and Marconi.
+
+[cols="1,1,1,1,1"]
+|===
+|Caracteristics | Solution 1 | Solution 2 | Solution 3 | Solution 4
+
+|Number of LedgerState computations | 2 | 1 | 1 | 1
+|Initial node resync required | No | Yes | Yes | Yes
+|Required node resync after Marconi logic change | No | No | Yes | Yes
+|Change to cardano-node | No | Yes | No | No
+|===
+
+[cols="1,1"]
+|===
+|Caracteristics | Description
+
+|Number of LedgerState computations
+|As it suggest, corresponds to total number of independent computations of the LedgerState.
+
+|Initial node resync
+|Yes if a resync of the relay node is required. No otherwise.
+
+|Node resync after schema change
+|Yes if a resync of the relay node is required after any changes to the indexing logic of Marconi.
+
+|Change to cardano-node
+|Yes if the solution requires a change to the cardano-node codebase.
+
+|===
+
+For more details, read the sections below.
+
+=== 1. Compute SDD by computing the `LedgerState`
+
+This is the approach that `cardano-db-sync` currently uses to compute the SDD for each epoch.
+
+A container diagram of this solution would look like:
+
+image::http://www.plantuml.com/plantuml/png/jL0_Q-Cm5D_zANIU4ZZ7SDnqKovkbz8w9g5pKAJNM4H_XFJKr9H-zqfAunfAe4ityNx_jVH87GObYrz2Cnaua1xH-dzrxUXfSH3OXttmu9ZH21eNpAZQIggJflB8fTZNpT-gIwV7X_jQKO_WucdNH0KL6jp2_eZ_3LXf1fL4EQ1e3OUPj-OtARVTcl0HB-VajIZ6FZEd8sK3aWuaQShv1DlR1AwaOKVIESC34_fmeQr4L6NvcHRBgg05emYUdEAMCXxRFosnpg1XHaRDn5lHD5D4piUhye83woCQYSbwPEIAJu2iZywPKVb1CmBFeD6VOvehCQDf-eZBc_otzXmonJyRCIfbzhtRhjjgqpNZt0dTNHdzjKtnLXJRmIEetUqBpY0h6-eurOPem9Dnnp4Xe-wx7xe4pOEIxm00[]
+
+Here is a general outline of the computation of the `LedgerState`:
+
+. At the start of `marconi-sidechain`, initialize the `LedgerState` (or `NewEpochState`) given the node genesis file
+. Get block from the chain sync protocol
+. Update the `LedgerState` given the block
+. Compute the SDD from the new LedgerState and store in the database
+. Repeat from step 2
+
+While this approach is the "easiest" to implement, it suffers from major drawbacks such as:
+
+* adding an additional 10-16GB of RAM in `marconi-sidechain` just from computing the `LedgerState`
+* slowing down indexing.
+  Computation of the `LedgerState` is _the_ bottleneck in `marconi-sidechain`
```

Comment:
Yes, AFAIU, when selecting a committee for the next epoch, they need to query all the indexers such as EpochSDD, so that would not help. The _real_ problem is the memory footprint though, which I mentioned above.